### ISSUE TITLE
Problems with dynamic filter predicates

### ISSUE THREAD

**Issue by beraadim** [jsonata-js/jsonata#100]
I am playing around with jsonata, and I like it _a lot_. However, I have run into a problem I do not understand. There might be something I have misunderstood. So my apologies if this is a false negative.

I tested this at http://try.jsonata.org/ (Not sure which version is being used on the site)

Concider the following JSON:
```
{
  "Contacts" : [
  {
    "type": "contact",
    "id": "69496314488018A264F30C62D6FE7E97",
    "firstName": "Name 1",
    "lastName": "Surname 1",
    "address1": "Adress 1",
    "address2": null,
    "zipCode": "Zip1",
    "city": "Country1",
    "countryCode": "NO",
    "phonePrivate": null,
    "phoneWork": null,
    "phoneMobile": "phone1",
    "fax": "",
    "email": "mail@mail1.com"
  },
  {
    "type": "contact",
    "id": "4742A63CF0B7F6861ACFC7A3543B9622",
    "firstName": "Name2",
    "lastName": "Surname2",
    "address1": "Adress2",
    "address2": null,
    "zipCode": "Zipcode2",
    "city": "City2",
    "countryCode": "NO",
    "phonePrivate": "phone2",
    "phoneWork": null,
    "phoneMobile": "mobile2",
    "fax": "fax2",
    "email": "mail@mail2.com"
   }
],
  
"Relations" : [
  {
    "apiId": "69496314488018A264F30C62D6FE7E97",
    "system": "SystemX",
    "systemId": "SystemX.Id.1"
  },
  {
    "apiId": "4742A63CF0B7F6861ACFC7A3543B9622",
    "system": "SystemX",
    "systemId": "SystemX.Id.2"
  }
 ]
}
```

I would think that the following jsonata would filter the relation tab from the contact:
```
$.{
	"Contacts" : $.Contacts[].{
        "ifsId" : $$.Relations[apiId = $.id],
        "id" : $.id
    }
}
```

But the result is this:
```
{
  "Contacts": [
    {
      "id": "69496314488018A264F30C62D6FE7E97"
    },
    {
      "id": "4742A63CF0B7F6861ACFC7A3543B9622"
    }
  ]
}
```

In other words, using the $.id does not result in a hit in the Relation tab as I would expect. If I replace $.id in the filter like this: Relations[apiId =  "4742A63CF0B7F6861ACFC7A3543B9622"], the Relations tab is filtered as expected. It seems that there is a problem using a dynamic value for the filtering.

**Comment by andrew-coleman** [comment]
The reason this is not doing what you expect is because the `$` context value within the `[apiId = $.id]` refers to each of the `Relations` array.  So you need to bind the `id` value to a variable before this context switch.  The following expression does this (http://try.jsonata.org/r1ZpBJGez):
```
{
  "Contacts" : $.Contacts[].(
    $id := id;
    {
      "ifsId" : $$.Relations[apiId = $id],
      "id" : $.id
    }
  )
}
``` 
I am trying to think of ways to make this type of 'cross referencing' easier, perhaps with some specific syntax that does the equivalent of an SQL JOIN.  But that's for the future.

**Comment by beraadim** [comment]
Thanks a lot! I am superhappy that what I tried is possible.  Closing this issue.
